### PR TITLE
Refactor framework

### PR DESCRIPTION
**Pull request by junjiejiangjjj**
**Refactor framework**

1. Remove task, add runners to run operators

   1. Use multiple runners (map / flatmap / generator / window and so on) to wrapper operators.

      The runner has a dead loop, read data, call `op()`, and write to the next dataframe.

      The runner has running status (running, failed, finished ...)

   2. Add `thread_pool_executor` to run the runner.

   3. Op ctx has multiple runners,  and it will call the runner's `join` function, wait for the runner end.

   4. Graph ctx wait for op ctx end by calling the `join` function.

   5. Remove multiple inputs.

   6. A better way to deal with errors.

   

2. Add map / filter / flatmap / generator runners

### CODE REVIEW

**Comment by filip-halt** [comment]
Should we also remove _operator_io.py with these changes? Seems like init, reader, and writer now contain all the necessary code.

File: towhee/__init__.py [hunk 1]
```diff
@@ -44,55 +45,86 @@ class _PipelineWrapper:
...
-            inputs = [args]
+        vargs = []
+        for arg in args:
+            vtype = type(arg).__name__
+            vargs.append(Variable(vtype, arg))
+        vargs = tuple(vargs)
 
+        # Process the data through the pipeline.
         in_df = DataFrame('_in_df')
-        for tup in inputs:
```
Comment (filip-halt) [inline towhee/__init__.py:72]: Are we purposely removing the ability to deal with multiple inputs?

File: towhee/engine/operator_io/_operator_io.py [hunk 2]
```diff
@@ -14,7 +14,7 @@
```
Comment (filip-halt) [inline towhee/engine/operator_io/_operator_io.py:14]: Do we still need this file, everything has moved to reader, init, and writer

Comment (junjiejiangjjj) [reply towhee/engine/operator_io/_operator_io.py:14]: We needn't this file, I forgot to delete it.

File: towhee/__init__.py (hunk 1 above)
Comment (junjiejiangjjj) [reply towhee/__init__.py:72]: Yeah, I think `pipeline()` is a very simple interface, users can send one input data, get the result if run success else raises an exception if have multiple inputs, users can write a loop and call `pipeline()` continually.

Comment (junjiejiangjjj) [reply towhee/__init__.py:72]: We will support `data_source` op, such as `local_data_source, `hdfs_data_source`, and so on, these can provide continuous input data.

**Comment by fzliu** [comment]
Can you refactor this into ~3 commits?

**Comment by fzliu** [comment]
Aren't we upgrading to `dataframe_v2` in this PR?

**Comment by junjiejiangjjj** [comment]
> Aren't we upgrading to `dataframe_v2` in this PR?

`dataframe_v2`  is not finished, we may implement it next version.

**Comment by fzliu** [comment]
Merging this PR first, @junjiejiangjjj will submit another PR if further changes are necessary.

File: towhee/dataframe/dataframe.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ def __init__(self, name: str = None, cols=None, data: List[Tuple[Variable]] = No
         self._sealed = False
         self._lock = threading.Lock()
         self._seal_cv = threading.Condition(self._lock)
+        self._accessible_cv = threading.Condition(self._lock)
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply towhee/dataframe/dataframe.py:59]: It's used to block the `get` function.

File: towhee/engine/operator_context.py [hunk 4]
```diff
@@ -12,18 +12,26 @@
...
+from towhee.engine.thread_pool_task_executor import ThreadPoolExecutor
 
 
-class OperatorContext(HandlerMixin):
+class OpStatus(Enum):
+    NOT_RUNNING = auto()
+    RUNNING = auto()
+    FINISHED = auto()
+    FAILED = auto()
+    STOPPED = auto()
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply towhee/engine/operator_context.py:31]: I forgot to delete it.

File: towhee/engine/engine.py [hunk 5]
```diff
@@ -73,13 +73,16 @@ def __init__(self):
...
+        self._setup_execs()
+        self._setup_sched()
 
     def stop(self) -> None:
-        self._task_sched.stop()
-        for task_exec in self._task_execs:
-            task_exec.stop()
+        self._sched.stop()
+        for executor in self._execs:
+            executor.stop()
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply towhee/engine/engine.py:82]: Executor's `stop()`  just stop to start new runners, it needn't  `join()`. The scheduler's `join()`  will wait for all ops end.

**Pull request merged**